arangod/Aql/HashedCollectExecutor.cpp

```diff
@@ -205,65 +203,109 @@ void HashedCollectExecutor::writeCurrentGroupToOutput(OutputAqlItemRow& output)
+  return ExecutorState::DONE;
 }
 
-std::pair<ExecutionState, NoStats> HashedCollectExecutor::produceRows(OutputAqlItemRow& output) {
+/**
+ * @brief Produce rows.
+ *   We need to consume all rows from the inputRange, except the
+ *   last Row. This is to indicate that this executor is not yet done.
+ *   Afterwards we write all groups into the output.
+ *   Only if we have written the last group we consume
+ *   the remaining inputRow. This is to indicate that
+ *   this executor cannot produce anymore.
+ *
+ * @param inputRange Data from input
+ * @param output Where to write the output
+ * @return std::tuple<ExecutorState, NoStats, AqlCall>
+ */
+
+auto HashedCollectExecutor::produceRows(AqlItemBlockInputRange& inputRange,
+                                        OutputAqlItemRow& output)
+    -> std::tuple<ExecutorState, NoStats, AqlCall> {
   TRI_IF_FAILURE("HashedCollectExecutor::produceRows") {
     THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
   }
-
   if (!_isInitialized) {
-    // fetch & consume all input and initialize output cursor
-    ExecutionState state = init();
-    if (state == ExecutionState::WAITING) {
-      return {state, NoStats{}};
+    // Consume the input range
+    _isInitialized = consumeInputRange(inputRange);
+  }
+
+  if (_isInitialized) {
+    while (_currentGroup != _allGroups.end() && !output.isFull()) {
+      writeCurrentGroupToOutput(output);
+      ++_currentGroup;
+      output.advanceRow();
     }
-    TRI_ASSERT(state == ExecutionState::DONE);
-    _isInitialized = true;
   }
 
-  // produce output
-  if (_currentGroup != _allGroups.end()) {
-    writeCurrentGroupToOutput(output);
-    ++_currentGroup;
-    ++_returnedGroups;
-    TRI_ASSERT(_returnedGroups <= _allGroups.size());
+  AqlCall upstreamCall{};
+  // We cannot forward anything, no skip, no limit.
+  // Need to request all from upstream.
+  return {returnState(), NoStats{}, upstreamCall};
+}
+
+/**
+ * @brief Skip Rows
+ *   We need to consume all rows from the inputRange, except the
+ *   last Row. This is to indicate that this executor is not yet done.
+ *   Afterwards we skip all groups.
+ *   Only if we have skipped the last group we consume
+ *   the remaining inputRow. This is to indicate that
+ *   this executor cannot produce anymore.
+ *
+ * @param inputRange  Data from input
+ * @param call Call from client
+ * @return std::tuple<ExecutorState, NoStats, size_t, AqlCall>
+ */
+auto HashedCollectExecutor::skipRowsRange(AqlItemBlockInputRange& inputRange, AqlCall& call)
+    -> std::tuple<ExecutorState, NoStats, size_t, AqlCall> {
+  if (!_isInitialized) {
+    // Consume the input range
+    _isInitialized = consumeInputRange(inputRange);
   }
 
-  ExecutionState state = _currentGroup != _allGroups.end() ? ExecutionState::HASMORE
-                                                           : ExecutionState::DONE;
+  if (_isInitialized) {
+    while (_currentGroup != _allGroups.end() && call.needSkipMore()) {
+      ++_currentGroup;
```

Comment:
Could we iterate faster instead of +1 here?

Reply:
Potentially es, but this would require more fiddling with skip / fullCount numbers here that I would like to avoid.
Having all this groups created will be the performance killer.